<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{SNPClust applied to Europeans}
-->

# SNPClust 

SNPClust is demonstrated here on a small subset of the Human Genome Diversity Project (HGDP) dataset: 157 European samples out of 940 and 5,000 SNPs out of 660,918.

```{r, message = FALSE, warning = FALSE}
library(snpclust)
library(magrittr)
```

The HGDP files are first converted to the snpgds format (cf. R package SNPRelate).

```{r, message = FALSE, warning = FALSE}
gds_path <- save_hgdp_as_gds()
```

SNPClust is then called on the GDS filepath.

```{r, message = FALSE, warning = FALSE}
snpclust_object <- snpclust(gds = gds_path, n_axes = 20)
```
```{r, echo = FALSE, results = 'hide'}
file.remove(gds_path)
```

# Quality control

Details about the quality control of the dataset are stored in a data frame.

```{r, results = 'asis'}
knitr::kable(snpclust_object$qc, 'markdown')
```

# Principal Component Analysis (PCA)

The results of PCA applied to the quality controlled dataset are stored in a long data frame.
Here we see that samples are grouped by country of origin.

```{r}
ggplot_pca(pca = snpclust_object$pca, groups = 'population', ellipses = TRUE)
plot_pca_pairs(seq_axes = 1:3, pca = snpclust_object$pca, groups = 'population')
plot_pca_pairs(seq_axes = 4:6, pca = snpclust_object$pca, groups = 'population')
```

# SNPs contributions to Principal Components (PCs)

For each PC, the absolute SNP contributions are displayed. SNPs are displayed by chromosome and position.

```{r}
ggplot_manhat(pca = snpclust_object$pca, gdata = snpclust_object$gdata) 
```

# SNPs selection by Gaussian mixture models

The Gaussian mixture models select SNPs above the background noise of other SNPs contributions.
Here the selected SNPs are colored in red.

```{r}
ggplot_selection(peaks = snpclust_object$peaks, pca = snpclust_object$pca) 
```

# PCA on the SNPClust selected dataset

When PCA is applied on the SNPClust selected dataset, samples are not grouped by geographic origin anymore.

```{r}
ggplot_pca(pca = snpclust_object$features_pca, groups = 'population', ellipses = TRUE)
plot_pca_pairs(seq_axes = 1:3, pca = snpclust_object$features_pca,
  groups = 'population')
plot_pca_pairs(seq_axes = 4:6, pca = snpclust_object$features_pca,
  groups = 'population')
```
